src/main/java/io/aiven/elasticsearch/gcs/GcsStoragePlugin.java

```diff
@@ -0,0 +1,86 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.elasticsearch.gcs;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import io.aiven.elasticsearch.gcs.storage.GcsBlobStoreRepository;
+import io.aiven.elasticsearch.storage.security.EncryptionKeyProvider;
+
+import org.elasticsearch.cluster.metadata.RepositoryMetaData;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.plugins.ReloadablePlugin;
+import org.elasticsearch.plugins.RepositoryPlugin;
+import org.elasticsearch.repositories.Repository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GcsStoragePlugin extends Plugin implements RepositoryPlugin, ReloadablePlugin {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GcsStoragePlugin.class);
+
+    private final GcsSettingsProvider gcsSettingsProvider;
+
+    public GcsStoragePlugin(final Settings settings) {
+        this.gcsSettingsProvider = new GcsSettingsProvider();
+        reload(settings);
+    }
+
+    @Override
+    public List<Setting<?>> getSettings() {
+        return List.of(
+                GcsStorageSettings.CREDENTIALS_FILE_SETTING,
+                GcsStorageSettings.PROJECT_ID,
+                GcsStorageSettings.CONNECTION_TIMEOUT,
+                GcsStorageSettings.READ_TIMEOUT,
+                EncryptionKeyProvider.PUBLIC_KEY_FILE,
+                EncryptionKeyProvider.PRIVATE_KEY_FILE
+        );
+    }
+
+    public Map<String, Repository.Factory> getRepositories(final Environment env,
+                                                           final NamedXContentRegistry namedXContentRegistry,
+                                                           final ClusterService clusterService) {
+        return Map.of(
+                GcsBlobStoreRepository.TYPE,
+                metadata -> createBlobStore(metadata, namedXContentRegistry, clusterService)
+        );
+    }
+
+    private GcsBlobStoreRepository createBlobStore(final RepositoryMetaData metadata,
+                                                   final NamedXContentRegistry namedXContentRegistry,
+                                                   final ClusterService clusterService) {
+        return new GcsBlobStoreRepository(metadata, namedXContentRegistry, clusterService, gcsSettingsProvider);
+    }
+
+    @Override
+    public void reload(final Settings settings) {
+        try {
+            gcsSettingsProvider.reload(settings);
+        } catch (final IOException e) {
+            LOGGER.error("Couldn't initialize GCS provider", e);
```

Comment:
From the documentation:
> Any failure during the operation should be signaled by raising an exception, but the plugin should otherwise continue to function unperturbed.

So I guess throwing `e` is necessary here.